packages/nx/src/project-graph/build-nodes/workspace-projects.ts

```diff
@@ -138,3 +145,16 @@ function normalizeProjectTargets(
   }
   return targets;
 }
+
+export function normalizeImplicitDependencies(
+  source: string,
+  implicitDependencies: ProjectConfiguration['implicitDependencies'],
+  context: ProjectGraphProcessorContext
+) {
+  return findMatchingProjects(
```

Comment:
we should rewrite this function. Currently the normalization of implicit dependencies is O(N^2), where N is the number of projects. In the worst case scenario it will be O(N^2) if say every project depends on every other project via '*', but his worst case scenario is unlikely.

What I would do is:
- Check if implicitDependencies is empty and skip the whole thing altogether.
- Check if there are some implicitDependencies using say '*', and if not, just assign them as project names.
- And only if there are implicitDeps that have special characters, I'd filter our the source and do the matching.

Reply:
Sure, I can add some of the short-circuiting logic.

Reply:
I'm pushing up the changes here now. To summarize, this is the approach I took:
- Updated `findMatchingProjects` to short circuit if the nameOrPattern doesn't contain special characters. In this case, it will add it if its a valid project, and skip the glob matching entirely regardless.
- If implicit dependencies is null or an empty array, we don't call into findMatchingProjects at all and simply return an empty array.
- We pass in the list of projectNames to `normalizeImplicitDependencies`. This should be the biggest boost in perf compared to the last pass at this. Now we find any matching projects, and then filter the source project out of the matches. This should mean filtering a much smaller array for large workspaces.